predicators/perception/spot_bike_perceiver.py

```diff
@@ -1,67 +1,104 @@
+            self._curr_env, SpotBikeEnv)
+        controller_name, objects, _ = self._curr_env.parse_action(action)
+        # The robot is always the 0th argument of an
+        # operator!
+        if "grasp" in controller_name.lower():
+            assert self._holding_item_id_feature == 0.0
+            # We know that the object that we attempted to grasp was
+            # the second argument to the controller.
+            object_attempted_to_grasp = objects[1].name
+            grasp_obj_id = obj_name_to_apriltag_id[object_attempted_to_grasp]
+            self._holding_item_id_feature = grasp_obj_id
+        elif "place" in controller_name.lower():
+            self._holding_item_id_feature = 0.0
 
     def step(self, observation: Observation) -> State:
-        assert isinstance(observation, _PartialPerceptionState)
-        if self._prev_action is not None:
-            assert self._curr_env is not None and isinstance(
-                self._curr_env, SpotBikeEnv)
-            controller_name, objects, _ = self._curr_env.parse_action(
-                observation, self._prev_action)
-            # The robot is always the 0th argument of an
-            # operator!
-            spot = objects[0]
-            if "grasp" in controller_name.lower():
-                assert self._holding_item_id_feature == 0.0
-                # We know that the object that we attempted to grasp was
-                # the second argument to the controller.
-                object_attempted_to_grasp = objects[1].name
-                grasp_obj_id = obj_name_to_apriltag_id[
-                    object_attempted_to_grasp]
-                self._holding_item_id_feature = grasp_obj_id
-            elif "place" in controller_name.lower():
-                self._holding_item_id_feature = 0.0
-            observation.set(spot, "curr_held_item_id",
-                            self._holding_item_id_feature)
-        return observation
+        self._update_state_from_observation(observation)
+        return self._create_state()
+
+    def _update_state_from_observation(self, observation: Observation) -> None:
+        assert isinstance(observation, _SpotObservation)
+        self._robot = observation.robot
+        self._known_object_poses.update(observation.objects_in_view)
+        self._nonpercept_atoms = observation.nonpercept_atoms
+        self._nonpercept_predicates = observation.nonpercept_predicates
+        self._gripper_open_percentage = observation.gripper_open_percentage
+
+    def _create_state(self) -> _PartialPerceptionState:
+        # Build the continuous part of the state.
+        assert self._robot is not None
+        state_dict = {
+            self._robot: {
+                "gripper_open_percentage": self._gripper_open_percentage,
+                "curr_held_item_id": self._holding_item_id_feature,
+                # Coming soon
+                "x": -100,
+                "y": -100,
+                "z": -100,
```

Comment:
do we not currently have access to this? We should open a PR ASAP if not, so we can implement a `Reachable` classifier!

Reply:
yeah we probably do, I just didn't look into it yet. agreed!